framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -949,146 +951,175 @@ public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
     }
 
     /**
-     * Checks all the preconditions of the method invocation or variable access {@code tree} with
-     * element {@code invokedElement}.
+     * Checks all the preconditions of the method invocation {@code tree} with
+     * element {@code invokedMethodElement}.
+     *
+     * @param methodInvocationTree the MethodInvocationTree to check preconditions for.
+     * @param invokedMethodElement the ExecutableElement to retrieve the method preconditions from.
      */
-    protected void checkPreconditions(Tree tree,
-            Element invokedElement, boolean methodCall) {
-        checkPreconditions(tree, invokedElement, methodCall, null);
+    protected void checkMethodInvocationPreconditions(MethodInvocationTree methodInvocationTree,
+            ExecutableElement invokedMethodElement) {
+        if (invokedMethodElement == null) {
+            return;
+        }
+
+        checkPreconditions(methodInvocationTree, contractsUtils.getPreconditions(invokedMethodElement));
     }
 
     /**
-     * Checks all the preconditions of the method invocation or variable access {@code tree} with
-     * element {@code invokedElement}.
+     * Checks that all the given {@code preconditions} hold true immediately prior to
+     * the method invocation or variable access at {@code tree}.
+     *
+     * @param tree the Tree immediately prior to which the preconditions must hold true.
+     * @param preconditions the preconditions to be checked.
      */
-    protected void checkPreconditions(Tree tree,
-            Element invokedElement, boolean methodCall, Set<Pair<String, String>> additionalPreconditions) {
-        Set<Pair<String, String>> preconditions = invokedElement == null ?
-                new HashSet<Pair<String, String>>() :
-                contractsUtils.getPreconditions(invokedElement);
+    protected void checkPreconditions(Tree tree, Set<Pair<String, String>> preconditions) {
+        // This check is needed for the GUI effects and Units Checkers tests to pass.
+        // TODO: Remove this check and investigate the root cause.
+        if (preconditions.isEmpty()) {
+            return;
+        }
 
-        if (additionalPreconditions != null) {
-            preconditions.addAll(additionalPreconditions);
+        checkPreconditions(tree, atypeFactory.getNodeForTree(tree), preconditions);
```

Comment:
I find the granularity of the different methods a bit strange. The previous only computed the preconditions for an element, this one computes the Node for the Tree. Do you need both? Can they be used interchangeably?
I would prefer eventually introducing a nicer abstraction instead of the `Pair<String, String>` which doesn't provide any type safety.